File: app/core/Tips.hs
module Tips (giveTip) where

import Card
import Bot
import State


-- mão do jogador -> deck a IA
giveTip :: BattleState -> String
giveTip battlestate 
    | score <= 20 = tipBotWillPlay battlestate
    | score <= 40 = tipToPlay battlestate
    | otherwise   = tipNotToPlay battlestate
    where score = playerScore battlestate

tipBotWillPlay :: BattleState -> String
tipBotWillPlay battlestate = "O bot talvez jogue: " ++ tipBotAux battlestate

tipBotAux :: BattleState -> String
tipBotAux battlestate = do
    let botDeck = cpuDeck battlestate
    auxTranslate $ element (botDeck !! makeChoice battlestate 10)


tipToPlay :: BattleState -> String
tipToPlay battlestate = do
    let seed = cpuScore battlestate
    let weight = wDeck (cpuDeck battlestate) 10
    let player_hand = take 5 (playerDeck battlestate)
    let w_player_hand = wHand player_hand weight
    let len = length w_player_hand

    "Você deveria jogar: " ++ auxTranslate (element (player_hand !! (w_player_hand !! nRandom seed (len - 1))))

tipNotToPlay :: BattleState -> String
tipNotToPlay battlestate = do 
    let botplay = tipBotAux battlestate
    case botplay of
        "Fogo" -> tipNotAux "Natureza"
        "Natureza" -> tipNotAux "Água"
        "Água" -> tipNotAux "Metal"
        "Metal" -> tipNotAux "Terra"
        "Terra" -> tipNotAux "Fogo"
        _ -> "Fogo"
    
tipNotAux :: String -> String
tipNotAux tipo = "Você não deveria jogar: " ++ tipo

-- | Retorna o tipo de elemento correspondente ao tipo definido na carta.
auxTranslate :: String -> String
auxTranslate word
    | word == "fire" = "Fogo"
    | word == "nature" = "Natureza"
    | word == "water" = "Água"
    | word == "metal" = "Metal"
    | otherwise = "Terra"

File: app/core/Tips.pl
:- consult('../util/StateManager.pl').
:- consult('../util/Helpers.pl').
:- consult('./Bot.pl').

/*
 * Interface que fornece a dica ao jogador.
 */
giveTip(Tip):-
    get_player_state(PlayerState),
    nth0(0, PlayerState, Score),
    selectTip(Score, Tip).

/*
 * Seleciona o tipo de dica, de acordo com a pontuação
 * atual da partida.
 */
selectTip(Score, Tip):-
    Score =< 20,
    botSelection(Card),
    tipBotWillPlay(Card, Tip),
    !.
selectTip(Score, Tip):-
    Score =< 40,
    tipToPlay(Tip),
    !.
selectTip(_, Tip):-
    botSelection(Card),
    tipNotToPlay(Card, Tip),
    !.

/*
 * Utiliza o mecanismo de seleção de carta do bot para predizer, qual carta
 * o bot tem probabilidade de jogar durante a rodada.
 */
tipBotWillPlay(Card, Tip):-
    get_elem(Card, Element),
    translate_elem(Element, TElement),
    string_concat("O bot talvez jogue: ", TElement, Tip).

/*
 * Utiliza o mecanismo de seleção de carta do bot para predizer, com a mão
 * do jogador, qual carta tem mais chance de vencer a rodada.
 */
tipToPlay(Tip):-
    makePossibleChoice(10, Card),
    get_elem(Card, Element),
    translate_elem(Element, TElement),
    string_concat("Você deveria jogar: ", TElement, Tip).

/*
 * Baseada na carta sorteada pelo bot, informa ao jogador qual carta ele deveria 
 * evitar jogar durante a rodada.
 */
tipNotToPlay(Card, Tip):-
    get_elem(Card, Element),
    notPlay(Element, ElementTip),
    translate_elem(ElementTip, TElement),
    string_concat("Você não deveria jogar: ", TElement, Tip).

/*
 * Fatos auxiliares a regra tipNotToPlay
 */
notPlay("fire", "nature").
notPlay("nature", "water").
notPlay("water", "metal").
notPlay("metal", "earth").
notPlay("earth", "fire").

/*
 * Chama o mecanismo de escolha do bot para selecionar uma carta
 * utilizando o nível de dificuldade 10.
 */
botSelection(Card):-
    get_bot_state(BotState),
    nth0(2, BotState, BotDeck),
    slice(0, 5, BotDeck, BotHand),
    makeChoice(10, CardIndex),
    get_card(CardIndex, BotHand, Card).

/*
 * Adaptação do mecanismo de seleção de carta do bot para selecionar
 * uma carta do jogador, que tem mais chance de vencer o bot.
 */
makePossibleChoice(Level, Card):-
    get_player_state(PlayerState),
    nth0(2,PlayerState,Deck),
    slice(0, 5, Deck, Hand),
    get_bot_state(BotState),
    wDeck(Level,BotState,Weight),
    wHand(Hand,Weight, W_player_hand),
    length(W_player_hand, L),
    random(0,L,Choice),
    nth0(Choice,W_player_hand,Id),
    nth0(Id,Hand,Card).




